### PR TITLE
feat(valid-expect): support using typechecking to ensure `toThrow` type assertions are passed a function

### LINKED ISSUE #1329: Ensure expect(function).toThrow()
When is possible to determine if the expect(value) is not a function when used with .toThrow(), it must be an error.

```ts
const str = "any value";

expect(str).toThrow() // Not works.
```

```ts
const returnStr = () => "any value";

expect(returnStr).toThrow() // Not handle, but is a valid case (if returnStr() throws an exception).
```

### PR DESCRIPTION
**Pull request by G-Rath**
Resolves #1329

### CODE REVIEW

File: src/rules/valid-expect.ts [hunk 1]
```diff
@@ -337,6 +355,30 @@ export default createRule<[Options], MessageIds>({
...
+            'toThrow',
+            'toThrowError',
+            'toThrowErrorMatchingSnapshot',
+            'toThrowErrorMatchingInlineSnapshot',
+          ].includes(getAccessorValue(matcher))
+        ) {
+          if (
+            services.getTypeAtLocation(expect.arguments[0]).getCallSignatures()
+              .length === 0
+          ) {
```
[the remark this answers is not in the document]
Comment (G-Rath) [reply src/rules/valid-expect.ts]: That does not seem to be true - as far as I can tell, TS does not explicitly type `any` as callable at the AST level, and I don't think it's a guaranteed false positive because by definition it could be "anything" so it could be a false positive or negative.

I think it's fair for us to be reporting on both `any` and `unknown` because of that + the solution being to just case to a callable value; the only edge case is with fixing, since doing that might actually break tests 🤔

File: docs/rules/valid-expect.md [hunk 2]
```diff
@@ -71,6 +75,14 @@ This rule is enabled by default.
 }
 ```
 
+### `typecheck`
+
+Use TypeScript type information to check that `toThrow` type assertions are
+being passed a callable.
```
[the remark this answers is not in the document]
Comment (G-Rath) [reply docs/rules/valid-expect.md]: That's what the "`toThrow` type assertions" means
